packages/bigquery/src/index.js

```diff
@@ -426,6 +516,89 @@ BigQuery.prototype.createDataset = function(id, options, callback) {
 BigQuery.prototype.createQueryStream = common.paginator.streamify('query');
 
 /**
+ * Creates a job. Typically when creating a job you'll have a very specific task
+ * in mind. For this we recommend one of the following methods:
+ *
+ * - {module:bigquery#startQuery}
+ * - {module:bigquery/table#copy}
+ * - {module:bigquery/table#copyFrom}
+ * - {module:bigquery/table#export}
+ * - {module:bigquery/table#import}
+ *
+ * However in the event you need a finer level of control over the job creation,
+ * you can use this method to pass in a raw [Job resource](https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs)
+ * object.
+ *
+ * @resource [Jobs Overview]{@link https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs}
+ * @resource [Jobs: insert API Documentation]{@link https://cloud.google.com/bigquery/docs/reference/v2/jobs/insert}
+ *
+ * @param {object} options - Object in the form of a [Job resource](https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs);
+ * @param {string=} options.jobPrefix - Prefix to apply to the job id.
+ * @param {function} callback - The callback function.
+ * @param {?error} callback.err - An error returned while making this request.
+ * @param {module:bigquery/job} callback.job - The newly created job.
+ * @param {object} callback.apiResponse - The full API response.
+ *
+ * @example
+ * var options = {
+ *   configuration: {
+ *     query: {
+ *       query: 'SELECT url FROM [publicdata:samples.github_nested] LIMIT 100'
+ *     }
+ *   }
+ * };
+ *
+ * bigquery.createJob(options, function(err, job) {
+ *   if (err) {
+ *     // Error handling omitted.
+ *   }
+ *
+ *   job.getQueryResults(function(err, rows) {});
```

Comment:
Why is `getQueryResults` being called here? If it was meant to wait, shouldn't it poll?

Reply:
Internally we recursively call `getQueryResults`, should we be waiting to start this process?

Reply:
`getQueryResults` should only be called for query jobs. I don't see a check here for the the type of job. When I try to call `getQueryResults` on a copy job resource, I get an error response

```
{
 "error": {
  "errors": [
   {
    "domain": "global",
    "reason": "invalid",
    "message": "Job 769108155013:8a432e18-336e-489f-bb90-834327622456 is not a query job",
   }
  ],
  "code": 400,
  "message": "Job 769108155013:8a432e18-336e-489f-bb90-834327622456 is not a query job"
 }
}
```

The more general way to check for job completion is to repeatedly [get the job resource](https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs/get) (note: does not wait for a timeout like `getQueryResults` does, so a polling interval is needed) and check for the [status.state](https://cloud.google.com/bigquery/docs/reference/rest/v2/jobs#status.state) `DONE`.

Reply:
I don't think it's possible to reliably sniff for this. If a user knows the job id, they could do something like this

```js
const job = bigquery.job('my-job-id')
const rows = await job.getQueryResults()
```

And we would have no way to detect what kind of job it actually is without an additional request. I think our best bet is to allow the upstream API to do the error handling and improve our documentation on when to call `get` and when to call `getQueryResults`

Reply:
What do you mean? This function is called `createJob`. You should be able to check for the type of job from the options object.

If it was `createQueryJob` then by all means call `getQueryResults`. But it's no good to call that for all jobs.

Reply:
When creating a job it's possible, but if a user were to simply reference a job id, then it would be impossible. At that point we're handling the same error in a different way. For the sake of consistency I'd prefer to let the upstream API handle it

Reply:
Maybe I'm missing something? The line I'm referring to is in a generic createJob function, which should not be calling `getQueryResults` since it only applies to query jobs.

Reply:
This comment chain seems to be in reference to our example code in the documentation comment block. The full code of the example snippet:

```js
 * var options = {
 *   configuration: {
 *     query: {
 *       query: 'SELECT url FROM [publicdata:samples.github_nested] LIMIT 100'
 *     }
 *   }
 * };
 *
 * bigquery.createJob(options, function(err, job) {
 *   if (err) {
 *     // Error handling omitted.
 *   }
 *
 *   job.getQueryResults(function(err, rows) {});
 * });
```

So this example shows how to use `createJob()` to create a query job, and use the returned `job` object's `getQueryResults()` to get the results.

The actual code of the method `createJob()` is here: https://github.com/callmehiphop/gcloud-node/blob/af1fc347f8a283faf13fdae0a1cb7f3cdddb3d93/packages/bigquery/src/index.js#L568

Sorry if I'm misunderstanding.

Reply:
🤦‍♂️ You're right. This example code makes sense. I should have realized it was example code, sorry.